image_segmentation/pytorch/runtime/training.py

```diff
@@ -24,104 +26,121 @@ def get_optimizer(params, flags):
     return optim
 
 
-def lr_warmup(optimizer, init_lr, lr, current_epoch, warmup_epochs):
-    scale = current_epoch / warmup_epochs
+def lr_warmup(optimizer, init_lr, lr, current_samples, warmup_samples):
+    scale = min(current_samples / warmup_samples, 1.0)
     for param_group in optimizer.param_groups:
         param_group['lr'] = init_lr + (lr - init_lr) * scale
 
 
 def train(flags, model, train_loader, val_loader, loss_fn, score_fn, device, callbacks, is_distributed):
     rank = get_rank()
     world_size = get_world_size()
+    global_batch_size = world_size * flags.batch_size * flags.ga_steps
+    next_eval_at = flags.start_eval_at
+
     torch.backends.cudnn.benchmark = flags.cudnn_benchmark
     torch.backends.cudnn.deterministic = flags.cudnn_deterministic
 
     optimizer = get_optimizer(model.parameters(), flags)
-    if flags.lr_decay_epochs:
+    if flags.lr_decay_samples:
         scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer,
-                                                         milestones=flags.lr_decay_epochs,
+                                                         milestones=flags.lr_decay_samples,
                                                          gamma=flags.lr_decay_factor)
     scaler = GradScaler()
 
     model.to(device)
     loss_fn.to(device)
-    if flags.normalization == "syncbatchnorm":
-        model = torch.nn.SyncBatchNorm.convert_sync_batchnorm(model)
     if is_distributed:
         model = torch.nn.parallel.DistributedDataParallel(model,
                                                           device_ids=[flags.local_rank],
                                                           output_device=flags.local_rank)
 
     stop_training = False
+    seen_samples = 0
+    model.train()
+    for callback in callbacks:
+        callback.on_fit_start()
     for epoch in range(1, flags.epochs + 1):
         cumulative_loss = []
-        if epoch <= flags.lr_warmup_epochs and flags.lr_warmup_epochs > 0:
-            lr_warmup(optimizer, flags.init_learning_rate, flags.learning_rate, epoch, flags.lr_warmup_epochs)
+        if seen_samples <= flags.lr_warmup_samples + TRAIN_DATASET_SIZE and flags.lr_warmup_samples > 0:
```

Comment:
Mehmet, thanks for your answer. I'll have to think about it and try few things, but it won't happen this week. I'll try to get back to you early next week.

But in the meantime, I had a look at the doc you attached. Could you:
a) double-check whether your evaluation schedule? The first eval is at 168064 (correctly, after 168k samples). The second eval should be at (168000+3360=) 171360 samples, so with your BS it should have taken place at 171392 (one row before the marked one). The following evals are as well lagging one row. Please let me know if this is an error or we have a misunderstanding here.
b) Add LR warmup schedule? There's only LR decay which I'd have to look at.

Thanks!

Reply:
Yes, you are right. I made couple mistakes for the eval points. I fixed them updated in this [doc](https://docs.google.com/spreadsheets/d/1LIF17aExHMUy6mHXxMIkhBV1D04TLqBXIJrBkvDPER4/edit?usp=sharing), I am assuming it should be good now. 

I also added an lr_warmup example here. I am assuming calculating lr warmup from either steps or samples are identical.

Reply:
Thanks for the update.

I'd rather be opposed to have functions dependent on the epochs. My intention was to remove them completely, but sadly it seems that `DistributedSampler`s need them. On the other hand, the RCP should be taken from a single-GPU run (with gradient accumulation if needed) rather than a multi-GPU run. Therefore, I'd rather "hack" the `DistributedSampler` and keep everything in samples, so:
`train_loader.sampler.set_epoch(samples_seen//TRAIN_DATASET_SIZE)`
and
`while samples_seen <= flags.max_samples:`
where `samples_seen` is updated every iteration. Does that make sense?

Now, two real questions remain
1. Where/when to update LR for LR warmup
a) After each iteration (which will cause some problems with logging)
b) After each "epoch". Sorry for using epochs again, but mlperf logger was not really designed for samples or iterations. If you are going with BS=128 then your `iteration == epoch` so it shouldn't matter. But I'd rather keep it after `TRAIN_DATASET_SIZE//BS` iterations.

2. Where/when to update LR for LR decay - the same story.